src/debugger/mock-ui/session.js

```diff
@@ -0,0 +1,239 @@
+      console.error(data.toString());
+      this.shutdown();
+    });
+  }
+
+  // callback to handle data written back by the adapter
+  _handleData(data: Buffer): void {
+    this._rawData = Buffer.concat([this._rawData, data]);
+    // the following code parses a message according to the protocol.
+    while (this._rawData.length > 0) {
+      // if we know what length we are expecting
+      if (this._contentLength >= 0) {
+        // we have enough data to check for the expected message
+        if (this._rawData.length >= this._contentLength) {
+          // first get the expected message
+          let message = this._rawData.toString("utf8", 0, this._contentLength);
+          // reduce the buffer by the message we got
+          this._rawData = this._rawData.slice(this._contentLength);
+          // reset the content length to ensure it is extracted for the next message
+          this._contentLength = -1;
+          // process the message
+          if (message.length > 0) {
+            try {
+              let msg = JSON.parse(message);
+              if (msg.type === "event") {
+                this._processEvent(msg);
+              } else if (msg.type === "response") {
+                this._processResponse(msg);
+              }
+            } catch (e) {
+              console.log("Invalid message");
+            }
+          }
+          continue; // there may be more complete messages to process
+        }
+      } else {
+        // if we don't know the length to expect, we need to extract it first
+        let idx = this._rawData.indexOf(TWO_CRLF);
+        if (idx !== -1) {
+          let header = this._rawData.toString("utf8", 0, idx);
+          let lines = header.split("\r\n");
+          for (let i = 0; i < lines.length; i++) {
+            let pair = lines[i].split(/: +/);
+            if (pair[0] === "Content-Length") {
+              this._contentLength = +pair[1];
+            }
+          }
+          this._rawData = this._rawData.slice(idx + TWO_CRLF.length);
+          continue;
+        }
+        // if we don't find the length we fall through and break
+      }
+      break;
+    }
+
+    //ask the user for the next command
+    this._reader.question("(dbg) ", (input: string) => {
+      this._dispatch(input);
+    });
+  }
+
+  _processEvent(event: DebugProtocol.Event) {
+    // to be implemented
+    console.log(event);
+  }
+
+  _processResponse(response: DebugProtocol.Response) {
+    // to be implemented
+    console.log(response);
+  }
+
+  // execute a command if it is valid
+  // returns whether the command was valid
+  _executeCommand(input: string): boolean {
+    let parts = input.split(" ");
+    let command = parts[0];
+
+    // for testing purposes, init and configDone are made into user commands
+    // they can be done without user input
+    if (command === "init") {
+      let args: DebugProtocol.InitializeRequestArguments = {
+        clientID: "CLI",
+        adapterID: "Prepack",
```

Comment:
Clearly this is not prescribed by the protocol. It does seem a terse. Could there in the future be another adapter that involves Prepack?